lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        # Register services with charm. This adds the events to observe.
+        self._systemd_notices = SystemdNotices(self, ["slurmd"])
+        event_handler_bindings = {
+            self.on.install: self._on_install,
+            self.on.stop: self._on_stop,
+            self.on.service_slurmd_started: self._on_slurmd_started,
+            self.on.service_slurmd_stopped: self._on_slurmd_stopped,
+        }
+        for event, handler in event_handler_bindings.items():
+            self.framework.observe(event, handler)
+
+    def _on_install(self, _: InstallEvent) -> None:
+        # Subscribe the charmed operator to the services on the machine.
+        # .subscribe() configures the notices hooks and starts the juju-systemd-notices daemon.
+        # The juju-systemd-notices daemon is per unit. This means that the unit name will be
+        # meshed into the service name. E.g. juju-systemd-notices becomes
+        # juju-{unit_name}-{unit_number}-systemd-notices.
+        self._systemd_notices.subscribe()
+
+    def _on_start(self, _: StartEvent) -> None:
+        # This will trigger the juju-systemd-notices daemon to
+        # emit a `service-slurmd-started` event.
+        systemd.service_start("slurmd")
+
+    def _on_stop(self, _: StopEvent) -> None:
+        # To stop the juju-systemd-notices service running in the background.
+        # .stop() also disables the juju-systemd-notices so that it does not
+        # start back up if the underlying machine is rebooted.
+        self._systemd_notices.stop()
+
+    def _on_slurmd_started(self, _: ServiceStartedEvent) -> None:
+        self.unit.status = ActiveStatus()
+        time.sleep(60)
+
+        # This will trigger the juju-systemd-notices daemon to
+        # emit a `service-slurmd-stopped` event.
+        systemd.service_stop("slurmd")
+
+    def _on_slurmd_stopped(self, _: ServiceStoppedEvent) -> None:
+        self.unit.status = BlockedStatus("slurmd not running")
+```
+"""
+
+__all__ = ["ServiceStartedEvent", "ServiceStoppedEvent", "SystemdNotices"]
+
+import argparse
+import asyncio
+import functools
+import logging
+import re
+import signal
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from types import MappingProxyType
+from typing import List, Union
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.constants import BusType, MessageType
+from dbus_fast.errors import DBusError
+from dbus_fast.message import Message
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = MappingProxyType(
+    {
+        "_5f": "_",
```

Comment:
Maybe just add an inline comment here: `# _ must be first`

Reply:
Good idea. I think I wrote this factoid down in one of my commit messages, but it would be good to also include this as an inline comment too for future reference.

Reply:
Inline comment added in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/a61cf98ddfaf54f73b3c56c9182208052ca01032